docs/rfcs/0002-storing-course-data.md

```diff
@@ -0,0 +1,96 @@
+## Storing course data
+
+### Abstract
+
+The aim of this RFC is to make decisions about the following:
+- The model definitions* that will be needed to store course data.
+- The plan for loading legacy course data, and keeping that data current
+  after we switch to using our own Open edX instance.
+
+Out of scope for this RFC:
+- Content management – we will just rely on Django admin for creating/updating
+  courses by hand.
+- Any kind of grading/certificate data.
+
+\* - _Much like MicroMasters, we will keep our own models for courses/programs
+  that will exist in parallel with the course data in edX_
+
+### Models
+
+Since this app has very similar needs, `Program`, `Course`, and `CourseRun` 
```

Comment:
How will this handle situations where a `Program` or `Course` spans multiple sources? It seems like those are likely to be varied and that `CourseRun` is (hopefully) the only entity where you won't have multiple records that represent the same logical entity. Do we need some kind of "merge key" in these records? I don't think `title` is a safe one to depend on for deduplicating these.

To illustrate, I could foresee the following scenario:

- `Program` named "Deep Learning" exists in both edX.org and Open edX
- Users took `Courses` A, B, and C with a mix of these happening on different platforms

I'd suggest for each of these tables we'd want three columns that we `unique_together`:
- `merge_key` - a string that represents a single logical entity (e.g. "Deep Learning Program") that may span multiple platforms (totally up for a better name too)
- `source` - what you already have here
- `external_id` - the id in the platform it's sourced from, same as the one you have on `CourseRun` here

We could then merge those records using a prioritized sort on `source`.

Reply:
My silent assumption was that we wouldn't need to 'join' programs or courses that moved platforms, but that may not be true. I think the `merge_key` idea is perfectly reasonable as a just-in-case. 

A couple things I'm not sure about:
- Putting `external_id` on all 3 tables. The notion of a program and a course (as opposed to a course run) does not really exist in edX, and as such there wouldn't be an id for them on the courseware platform.
- Adding `merge_key` to `CourseRun`. If a course changes platforms, I would think that we would need to create a new `Course` to represent it on the new platform. Assuming we do that, the `merge_key` of the `Course` would be all we need to relate `CourseRun`s across platforms

Reply:
If `Program` is an xPro-only entity, it shouldn't need a `source` field either then, right?

I'm fine on your second point, although there are benefits to storing that data on the both tables and avoiding an extra join/select if that's all you needed.

Reply:
I think only course run have external_id.